src/core/tranche/assessor.sol

```diff
@@ -0,0 +1,96 @@
+    function poolValue() public returns(uint);
+    function indexOf(address) public returns(int);
+    function operatorOf(uint) public returns(address);
+}
+
+contract Assessor {
+    // --- Data ---
+    TrancheManagerLike trancheManager; 
+
+    // --- Assessor ---
+    // computes the current asset value for tranches.
+    constructor(address trancheManager_) public {
+        trancheManager = TrancheManagerLike(trancheManager_);
+    }
+
+    function getAssetValueFor(address operator_) public returns (uint) {
+         OperatorLike operator = OperatorLike(operator_);
+         uint trancheDebt = operator.debt();
+         uint trancheReserve = operator.balance();
+         int trancheIndex = trancheManager.indexOf(operator_);
+
+         require(trancheIndex >= 0);
+
+         uint totalSeniorDebt = calcSeniorDebt(trancheIndex-1);    
+         uint totalEquityReserve = calcEquityReserve(uint(trancheIndex+1));
+         uint poolValue = trancheManager.poolValue();
+         if (trancheManager.isEquity(operator_)) {
+            return calcEquityAssetValue(poolValue, trancheReserve, totalSeniorDebt);
+         } 
+         return calcSeniorAssetValue(poolValue, trancheReserve, trancheDebt, totalEquityReserve, totalSeniorDebt);
+    }
+
+    // Tranche.assets (Equity) = (Pool.value + Tranche.reserve - Sum(Senior.debt)) > 0 && (Pool.value - Tranche.reserve - Sum(Senior.debt)) || 0
+    function calcEquityAssetValue(uint poolValue, uint trancheReserve, uint totalSeniorDebt) internal returns (uint) {
+        int assetValue = int(poolValue + trancheReserve - totalSeniorDebt);
+        return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+
+    // Tranche.assets (Senior) = (Tranche.debt < (Pool.value + Sum(Equity.reserve))) && (Sum(Senior.debt) + Tranche.reserve) || (Pool.value + Sum(Equity.reserve) + Tranche.reserve)
+    function calcSeniorAssetValue(uint poolValue, uint trancheReserve, uint trancheDebt, uint equityReserve, uint seniorDebt) internal returns (uint) {
+         // healthy: outstanding loans and / or reserves of tranches with higher risk can cover the tranche debt
+         if ((poolValue + equityReserve - seniorDebt ) >= trancheDebt) {
+            return trancheDebt + trancheReserve;
+         } 
+         // losses: tranche debt can not be covered anymore.
+         // Tranche assets contain everything left in the loan pool + reserves of tranches with higher risk minus the outstanding debt of tranches with lower risk 
+         int assetValue = int(poolValue + equityReserve + trancheReserve - seniorDebt);
+         return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+
+    // totalEquityReserve = sum of the reserves of all tranches with higher risk than the tranche with the index provided
+    function calcEquityReserve(uint startIndex) internal returns (uint) {
+        uint totalEquityReserve = 0;
+        for (uint i = startIndex; i < trancheManager.trancheCount(); i++) {
```

Comment:
i think this should be `trancheCount + 1` (index, not length)

Reply:
trancheCount is the length. The last element in the array has the position -> length-1